README.md

```diff
@@ -1,16 +1,204 @@
 
 ## Language
 
 The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
 
+# 1 Introduction
+
+## Separation
+
+Why separate the UCAN from the invocation format? Surely the UCAN itself already contains all the infomation required.
+
+I would argue that this is not the case for a few reasons.
+
+1. Authority is not intent to execute. Authority transfer is first-class in UCAN, other actions are not
+2. Mixing the two levels means that invocation info will live with the token (as a proof) on subdelegations
+3. Other detail, such as scheduling
+
+A job request may be picked up by anyone, PRIOR to UCAN delegation. A handshake or matchmaking may need to be performed.
+
+## Authority Is Not Intention
+
+Or to put it another way: "just because you can, doens't mean you should". Granting a UCAN to a peer means that they are allowed to perform some actions for a period of time. This is a feature not a bug, but also says nothing about the intention of when it should be run. I may grant a collaborative process the ability to perform actions on an ongoing basis (hmm, but vioating POLA)
+
+I could preload the service with `n` UCANs with narrow time windows and `nbf`s in the future. That would certainly be very secure, but it would be less convenient since I need to come online to issue more or to 
+
+# 2 Envelope
+
+UCAN uses a capabilities model. The 
+
+* UCANTO
+* zcap
+* IPVM
+
+
+
+
+Specify Roles
+
+
+
+
+## 2.1 IPLD
+
+``` ipldsch
+type Invocation struct {
+  inv [&UCAN] -- To invoke
+  v   SemVer  -- Version
+  nnc String  -- Nonce
+  sig Bytes   -- Signature
+}
+
+type SemVer struct {
+  ma Integer
+  mi Integer
+  pa Integer
+  ta optional nullable String
+}
+```
+
+## 2.2 Exmaple
+
+ ``` json
+{
+  "ucan/invoke": [ "bafyLeft", "bafyRight", "bafyEnd" ]
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "sig": 0xCOFFEE
+}
+```
+
+# 3 Receipt & Attestation
+
+An invocation receipt is a claim about what the output of an invocation is. A receipt MUST be attested via signature of the principal (the audience of the associated invocation).
+
+Note that this does not guarantee correctness of the result! The level of this statement's veracity MUST be ony taken that the signer claims this to be a fact.
+
+The receipt MUST be signed with by the `aud` from the UCAN.
+
+## 3.1 IPLD
+
+``` ipldsch
+type Receipt struct {
+  inv {&UCAN : {URI : Any}} <!-- not sure if this actually works? My guess is that the link doesn't work because it should not  -->
+  v   SemVer
+  nnc String
+  ext optional {String : Any}
+  sig Bytes
+}
+
+type URI = String
+```
+
+## 3.2 JSON Example
+
+``` json
+{
+  "ucan/receipt": {
+    "bafyLeft": {
+      "a": 42,
+      "example.com": {
+        "msg/read": [
+          "from": "alice@example.com",
+          "text": "hello world"
+        ]
+      }
```

Comment:
It is not clear what's the map with URL keys. I would have guessed a result object, but then not sure what the URLs are or why above one is capability and other is `42`.

Reply:
Sorry, there's not explanatory text yet — it's extremely WIP. Thanks for the early feedback, though!

The map is anything that the runner wants to attest from the invocation. The named value is the pure value returned from the invocation, but you're right that snce we're using UCANs, that could potentially be multiple resources together, so the pure value should be namespaced under e.g. the Wasm CID.